### PR TITLE
DHeap added

### LINKED ISSUE #102: Add m-ary heap
#### Description of the problem

Currently, the package does not have m-ary heap (a generalization of the binary heap).
#### References/Other comments
https://www.geeksforgeeks.org/k-ary-heap/

### PR DESCRIPTION
**Pull request by prshnt19**
#### References to other Issues or PRs or Relevant literature

Closes #102

### CODE REVIEW

**Comment by prshnt19** [comment]
The `_build(self)` in `BinaryHeap` doesn't seem to be doing anything. Its called once when the heap is to be built. At that time heap is empty so nothing actually happens.

**Comment by czgdp1807** [comment]
Well by default, elements can be `None` so `_build` does nothing. However, if you pass some list in elements then `_build` will work.

**Comment by czgdp1807** [comment]
Is it possible to remove the class `BinaryHeap` and add a constructor function with a same protoype as `BinaryHeap.__new__`?
```python
def BinaryHeap(arguments accepted by BinaryHeap.__new__)
```
We can also add a special constructor for `TernaryHeap`.

**Comment by prshnt19** [comment]
I think a better alternative will be to inherit `DHeap`. `BinaryHeap` and `TernaryHeap` will be a child class with d=2 and d=3 respectively.
```
class BinaryHeap(DHeap):
    def __new__(cls, elements=None, heap_property="min"):
         obj = DHeap.__new__(cls, elements=None, heap_property="min", d=2)
         return obj
```


```
class TernaryHeap(DHeap):
    def __new__(cls, elements=None, heap_property="min"):
         obj = DHeap.__new__(cls, elements=None, heap_property="min", d=3)
         return obj
```

**Comment by prshnt19** [comment]
ping @czgdp1807

**Comment by czgdp1807** [comment]
Can we use `DynamicOneDimensionalArray` [here](https://github.com/codezonediitj/pydatastructs/pull/118/files#diff-490d802c52ac67dcfaaf581599aaf045R86)?

File: pydatastructs/trees/heaps.py [hunk 1]
```diff
@@ -180,12 +182,131 @@ def __str__(self):
...
+    >>> max_heap.extract().key
+    7
+    >>> max_heap.insert(6, 6)
+    >>> max_heap.extract().key
+    6
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/D-ary_heap
```
Comment (czgdp1807) [inline pydatastructs/trees/heaps.py]: Any references for `TernaryHeap`?

File: pydatastructs/trees/heaps.py [hunk 2]
```diff
@@ -65,13 +67,14 @@ class BinaryHeap(Heap):
     References
     ==========
 
-    .. [1] https://en.m.wikipedia.org/wiki/Binary_heap
+    .. [1] https://en.wikipedia.org/wiki/D-ary_heap
     """
-    __slots__ = ['_comp', 'heap', 'heap_property', '_last_pos_filled']
+    __slots__ = ['_comp', 'heap', 'm', 'heap_property', '_last_pos_filled']
```
Comment (czgdp1807) [inline pydatastructs/trees/heaps.py]:
```suggestion
    __slots__ = ['_comp', 'heap', 'd', 'heap_property', '_last_pos_filled']

```
Let's keep `d` to denote the generalisation everywhere.

File: pydatastructs/trees/heaps.py [hunk 3]
```diff
@@ -180,12 +182,131 @@ def __str__(self):
...
+    >>> max_heap = TernaryHeap(heap_property='max')
+    >>> max_heap.insert(1, 1)
+    >>> max_heap.insert(5, 5)
+    >>> max_heap.insert(7, 7)
+    >>> max_heap.extract().key
+    7
+    >>> max_heap.insert(6, 6)
+    >>> max_heap.extract().key
+    6
+
```
Comment (czgdp1807) [inline pydatastructs/trees/heaps.py]: These examples do not show the working of `TernaryHeap`. The structure will be similar to `BinaryHeap`. May be one more insert would show the difference.

File: pydatastructs/trees/tests/test_heaps.py [hunk 4]
```diff
@@ -47,6 +47,88 @@ def test_BinaryHeap():
...
+    max_heap.insert(2, 2)
+    max_heap.insert(7, 7)
+    assert str(max_heap) == \
+           ('[(100, 100, [1, 2, 3]), (25, 25, [4, 5, 6]), '
+            '(36, 36, [7, 8]), (17, 17, []), '
+            '(3, 3, []), (19, 19, []), (1, 1, []), '
+            '(2, 2, []), (7, 7, [])]')
+
+    expected_extracted_element = max_heap.heap[0].key
+    assert max_heap.extract().key == expected_extracted_element
```
Comment (czgdp1807) [inline pydatastructs/trees/tests/test_heaps.py]:
```suggestion
    assert max_heap.extract().key == 100

```

File: pydatastructs/trees/tests/test_heaps.py [hunk 5]
```diff
@@ -47,6 +47,88 @@ def test_BinaryHeap():
...
+    for _ in range(8):
+        sorted_elements.append(max_heap.extract().key)
+    assert expected_sorted_elements == sorted_elements
+
+    elements = [
+        TreeNode(7, 7), TreeNode(25, 25), TreeNode(100, 100),
+        TreeNode(1, 1), TreeNode(2, 2), TreeNode(3, 3),
+        TreeNode(17, 17), TreeNode(19, 19), TreeNode(36, 36)
+    ]
+    min_heap = BinaryHeap(elements=elements, heap_property="min")
```
Comment (czgdp1807) [inline pydatastructs/trees/tests/test_heaps.py]:
```suggestion
    min_heap = TernaryHeap(elements=elements, heap_property="min")

```
May be you wanted the above.

File: pydatastructs/trees/tests/test_heaps.py [hunk 6]
```diff
@@ -47,6 +47,88 @@ def test_BinaryHeap():
...
+    max_heap = DHeap(heap_property="max", d=4)
+    assert max_heap.extract() is None
+    max_heap.insert(100, 100)
+    max_heap.insert(19, 19)
+    max_heap.insert(36, 36)
+    max_heap.insert(17, 17)
+    max_heap.insert(3, 3)
+    max_heap.insert(25, 25)
+    max_heap.insert(1, 1)
+    max_heap = DHeap(max_heap.heap, heap_property="max", d=3)
```
Comment (czgdp1807) [inline pydatastructs/trees/tests/test_heaps.py]: We should test it for `d=4` as `d=3` and `d=2` have already been tested.

File: pydatastructs/trees/heaps.py (hunk 2 above)
Comment (prshnt19) [reply pydatastructs/trees/heaps.py]: OK, it was by mistake.

File: pydatastructs/trees/heaps.py (hunk 3 above)
Comment (prshnt19) [reply pydatastructs/trees/heaps.py]: I can add one more insert. But one can't verify it's `TernaryHeap` heap by insertion unless we add feature like height. But adding feature like height is unnecessary.

File: pydatastructs/trees/heaps.py (hunk 1 above)
Comment (prshnt19) [reply pydatastructs/trees/heaps.py]: There was no individual reference to `TernaryHeap`. But it was there on the Wiki page of [DHeap](https://en.wikipedia.org/wiki/D-ary_heap)

File: pydatastructs/trees/tests/test_heaps.py (hunk 6 above)
Comment (prshnt19) [reply pydatastructs/trees/tests/test_heaps.py]: This line is actually for testing heap_property. 

The actual testing is on d=4 only. And d=3 is used for conversion from one type of heap to other.

**Comment by prshnt19** [comment]
> Can we use `DynamicOneDimensionalArray` [here](https://github.com/codezonediitj/pydatastructs/pull/118/files#diff-490d802c52ac67dcfaaf581599aaf045R86) instead of `list`?

The `DynamicOneDimensionalArray` itself uses `list`. I think using `DynamicOneDimensionalArray` would be a good idea only if we use `cyptes` for `DynamicOneDimensionalArray`. This will give a performance boost. I can create a new issue for that.

File: pydatastructs/trees/heaps.py (hunk 1 above)
Comment (prshnt19) [reply pydatastructs/trees/heaps.py]: I have updated the references.

**Comment by czgdp1807** [comment]
Though `DynamicOneDimensionalArray` uses list but the deletion time for the same is amortized O(1) with a nice trade off with space complexity. What happens if we use it?

File: pydatastructs/trees/heaps.py [hunk 7]
```diff
@@ -141,10 +143,10 @@ def insert(self, key, data):
         self.heap.append(new_node)
         self._last_pos_filled += 1
         i = self._last_pos_filled
-        self.heap[i].left, self.heap[i].right = 2*i + 1, 2*i + 2
+        self.heap[i].left, self.heap[i].right = self.d*i + 1, self.d*i + self.d
```
Comment (czgdp1807) [inline pydatastructs/trees/heaps.py]: May be, `left` -> `leftmost` and `right` -> `rightmost` will sound better.

Comment (prshnt19) [reply pydatastructs/trees/heaps.py]: But it may sounds odd for `BinaryHeap`.

Comment (czgdp1807) [reply pydatastructs/trees/heaps.py]: Two properties can be defined for `BinaryHeap`,
```python
@property
def left(self):
    return self.leftmost

@property
def right(self):
    return self.rightmost
```

Comment (prshnt19) [reply pydatastructs/trees/heaps.py]: ok

**Comment by prshnt19** [comment]
Used `DynamicOneDimensionalArray` instead of `List` in `DHeap`.  `DynamicOneDimensionalArray` could be made faster using `ctypes` array.

`ctypes` implementations are 10 to 100 times faster than python and takes less memory than python.

**Comment by czgdp1807** [comment]
> `DynamicOneDimensionalArray` could be made faster using `ctypes` array. `ctypes` implementations are 10 to 100 times faster than python and takes less memory than python.

Can you please make an experimental PR for the same?

File: pydatastructs/linear_data_structures/arrays.py [hunk 8]
```diff
@@ -255,6 +255,7 @@ def delete(self, idx):
     def size(self):
         return self._size
 
+
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/arrays.py]:
```suggestion
```

File: pydatastructs/trees/tests/test_heaps.py [hunk 9]
```diff
@@ -20,15 +21,16 @@ def test_BinaryHeap():
...
+        "(3, 3, []), (25, 25, []), (1, 1, []), "
+        "(2, 2, []), (7, 7, [])]")
 
-    expected_extracted_element = max_heap.heap[0].key
-    assert max_heap.extract().key == expected_extracted_element
+    assert max_heap.extract().key == 100
 
     expected_sorted_elements = [36, 25, 19, 17, 7, 3, 2, 1]
+    l=max_heap.heap[0].left
+    l=max_heap.heap[0].right
```
Comment (czgdp1807) [inline pydatastructs/trees/tests/test_heaps.py]:
```suggestion
    l = max_heap.heap[0].left
    l = max_heap.heap[0].right

```

**Pull request merged**